Add CRS.utm and support crs="utm" inputs in to_crs methods

Kirill888 opened the pull request:
## UTM convenience methods

1. `CRS.utm(some_geom)` picks appropriate utm zone for a given geometry and returns corresponding CRS object
2. When transforming geometry or geobox accept `crs="utm|utm-n|utm-s"` as a valid destination crs, meaning pick "appropriate" utm CRS for an object being transformed.
3. ~"appropriate UTM zone" is currently defined by the `lon,lat` of the centroid of the geometry~
4. UTM zones that overlap with `some_geom` are located using `pyproj`, then zone with biggest overlap is chosen


## Snap to

Adds `GeoBox.snap_to(other)`, this method adjust source geobox slightly by shifting footprint by a sub-pixel amount such that pixel edges align exactly with pixel edges of the `other` geobox. Method returns adjusted copy of self. Method only works when it is possible to align pixels by pure translation only - same projection, same resolution, same orientation. `ValueError` is raised when it's not possible.

Closes #59

## Enclosing

Related to `snap_to` somewhat. GeoBox construction method that creates a new "pixel grid compatible" geobox that minimally covers some area.

## Generalized Slicing into GeoBox

Overload `geobox[roi]` to support geometry types on input, footprint in the world can be converted to pixel space, from which pixel slice is computed.

Closes #31

Kirill888 committed:
Add CRS.utm constructor method

Pick appropriate UTM zone given some input and construct CRS for that
UTM zone.

Kirill888 committed:
Add support for crs="utm" in many places

- {Geometry|BoundingBox|GeoBox}.to_crs
- GeoBox.from_*
- GeoBox.footprint
- .odc.output_geobox

Kirill888 committed:
Spelling in docs

Kirill888 committed:
Implement GeoBox.snap_to #59

Kirill888 committed:
Add types to transformer cache

Also cache entire transformer and not just .transform function

snowman2 commented on odc/geo/crs.py:
```diff
@@ -309,6 +311,38 @@ def result(x, y):
         return ("odc.geo.crs.CRS", str(self))
 
+    @staticmethod
+    def utm(
```
I believe this method handles scenarios when the UTM zone is irregular:

https://pyproj4.github.io/pyproj/stable/examples.html#find-utm-crs-by-latitude-and-longitude

It is used by geopandas and rioxarray in their `estimate_utm_crs` methods.

snowman2 replied:
See: https://github.com/geopandas/geopandas/blob/1821b6d72791ce1f7da32db5b22e8803a3dc44b9/geopandas/array.py#L770

Kirill888 replied:
great, thanks for that, I'll change the logic to use pyproj query

Kirill888 committed:
Add BoundingBox.aoi property

simplify interop with pyproj.aoi

robbibt commented:
Hey @Kirill888, this looks fantastic - the UTM zone thing in particular looks really useful. One issue/question that occured to me however is that Landsat data is always packaged with UTM "North" zones (e.g. [EPSG:32656](https://epsg.io/32656)) even if the data itself lies in the "South" zone (e.g. [EPSG:32756](https://epsg.io/32756)) - if this was integrated into ODC-core and I tried native loading southern hemisphere data by supplying `crs="utm"`, would I risk having my data reprojected accidently? If so, I wonder if having an additional "utm-n" option could be useful, to force it to return the northern UTM zone even if the query is in the southern hemisphere (or maybe that's something that should be dealt with outside of `odc-geo`)

Kirill888 commented:
@robbibt pretty sure it would perform reproject when going from north utm to south utm.

So maybe `utm-n`, `utm-s` should be an option. Although I'm switching this code to use pyproj query functionality so I'll need to investigate best way to construct "same zone and datum UTM CRS, but in north hemisphere", maybe `pyproj` already exposes that. 

But what you really need for `.load` is `crs="native"` and `crs="auto"` where `native` raises an error when there are more candidates than 1 and `auto` just picks most common.

Kirill888 committed:
Use pyproj to find most appropriate UTM crs

This also adds an option to choose datum. UTM zone is chosen based on
the overlap between supplied geometry and valid region of the UTM zone.

Kirill888 committed:
Support crs=utm-n and crs=utm-s options

In some situations you want to ignore south/north distinction and always
use utm crs from a particular hemisphere.

Kirill888 committed:
Ensure that crs is None when requesting pixel mode

Kirill888 committed:
Add BoundingBox.round() method

Kirill888 committed:
Add GeoBox.project method

maps geometries between world and pixel coordinates, both ways.

Kirill888 committed:
Implement GeoBox.enclosing(region) method

Constructs same pixel grid geobox that covers supplied region

Kirill888 committed:
Implement bbox&bbox and bbox|bbox

Kirill888 committed:
Generalize GeoBox slicing to accept more types #31

Slice into geobox with types like Geometry|BoundingBox|GeoBox

Kirill888 committed:
Make pre-release 0.3.0rc0

snowman2 commented:
Side note: It may also be helpful to look for these projections as well:
- https://epsg.org/crs_32600/WGS-84-UTM-grid-system-northern-hemisphere.html
- https://epsg.org/crs_32700/WGS-84-UTM-grid-system-southern-hemisphere.html

Kirill888 commented:
> Side note: It may also be helpful to look for these projections as well:

hm, should we maybe use those when supplied region spans more than 2 UTM zones in the east-west direction?

ping: @robbibt, @SpacemanPaul
